File: plugins/collector/event/apps/service/nscd.go
package service

import (
	"collector/cache/process"
	"collector/event/apps"
	"regexp"
	"strings"
)

type Nscd struct {
	version string
	reg     *regexp.Regexp
}

func (Nscd) Name() string { return "nscd" }

func (Nscd) Type() string { return "service" }

func (n Nscd) Version() string { return n.version }

func (Nscd) Match(p *process.Process) bool { return p.Name == "nscd" }

func (n *Nscd) Run(p *process.Process) (m map[string]string, err error) {
	if n.reg == nil {
		n.reg = regexp.MustCompile(`nscd \(.*?\) (\d+\.)+\d+`)
	}
	result, err := apps.Execute(p, "--version")
	str := n.reg.FindString(result)
	if str == "" {
		err = apps.ErrVersionNotFound
		return
	}
	slice := strings.Split(result, " ")
	n.version = slice[len(slice)-1]
	return
}

func init() { apps.Regist(&Nscd{}) }

File: plugins/collector/src/event/apps/service/nscd.rs
use crate::event::apps::{execute, find_version, AppProc, IApp};
use std::collections::HashMap;

pub struct Nscd {
    version: String,
}

impl Nscd {
    pub fn new() -> Self { Nscd { version: String::new() } }
}

impl IApp for Nscd {
    fn name(&self)     -> &'static str { "nscd" }
    fn app_type(&self) -> &'static str { "service" }
    fn version(&self)  -> &str         { &self.version }

    fn matches(&mut self, p: &AppProc) -> bool { p.name == "nscd" }

    fn run(&mut self, p: &AppProc) -> anyhow::Result<HashMap<String, String>> {
        let out = execute(p, &["--version"])?;
        // "nscd (GNU C Library) 2.35 …"
        self.version = find_version(&out);
        Ok(HashMap::new())
    }
}
